fedbiomed/transport/node_agent.py

```diff
@@ -112,11 +155,14 @@ async def send(self, message: Message) -> None:
                             "this node to convert it as ACTIVE. Node will be updated "
                             "as DISCONNECTED soon if no request received.")
 
-        try:
-            await self._queue.put(message)
-        except Exception as exp:
-            raise FedbiomedCommunicationError(
-                f"{ErrorNumbers.FB628}: Can't send message to the client. Exception: {exp}")
+        # Updates replies
+        async with self._replies_lock:
+            self._replies.update({
```

Comment:
Hi @srcansiz 
I wonder if we should add:
```
if message.request_id:
   self._replies.update(....)
```

This avoids running the `on_reply` if we send a `message` with a `request_id` of `None` or empty string.
Though this case would be a programmer error, this would be pretty annoying. For example, with `request_id = None` it would divert a LOG or SCALAR message, etc.
This case is not covered/secured by the type check in `GrpcServer.send()` which checks that message is ` Message` (that's fine).
Currently this case cannot happen as in current version, `GrpcServer.send()` is called only from `Request.send()` which adds a `request_id`.

Another option would be to add a check in `GrpcServer.send()` that `request_id` is not `None` / not empty string.

This would make sense for the sake of modularization between gRPC layer / upper layers.

What do you think ?

Reply:
Hi Marc, 

`LOG` and `SCALAR` messages has their own channel and service so request replies never confused with `Feedback` channel. If `request_id` is `None`. `NodeAgent` will create a key in `self._replies` as `None` like `{None: {on_reply: callback}}`. This can be dangerous because developer won't see any error. Therefore, I agree that we may add your modification.  

I was actually thinking about removing `GrpcServer.send` because it is not used anymore. All messages are send through `NodeAgent.send`. It is because each message from researcher to node is considered as request. I would suggest to remove it for now and add it when there is use-case for it.

Reply:
> LOG and SCALAR messages has their own channel and service so request replies never confused with Feedback channel
You're right, thanks for clarifying.

> Therefore, I agree that we may add your modification.
OK.
Maybe I would favor implementation which checks in GrpcServer.send() that request_id is not None / not empty string (to have all checks in same place) - but it's up to you.

> I was actually thinking about removing GrpcServer.send
OK for me

Reply:
Discussion: only test in `on_reply`, not `GrpcServer.send()` (because it's ok to send messages without request_id